src/webgpu/api/operation/memory_sync/buffer/rw_and_wr.spec.ts

```diff
@@ -16,7 +16,101 @@ TODO: Tests with more than one buffer to try to stress implementations a little
 `;
 
 import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import {
+  kOperationBoundaries,
+  kBoundaryInfo,
+  OperationContextHelper,
+} from '../operation_context_helper.js';
 
-import { BufferSyncTest } from './buffer_sync_test.js';
+import {
+  kAllReadOps,
+  kAllWriteOps,
+  BufferSyncTest,
+  checkOpsValidForContext,
+} from './buffer_sync_test.js';
 
 export const g = makeTestGroup(BufferSyncTest);
+
+g.test('rw')
+  .desc(
+    `
+    Perform a 'read' operations on a texture subresource, followed by a 'write' operation.
+    Operations are separated by a 'boundary' (pass, encoder, queue-op, etc.).
+    Test that the results are synchronized.
+    The read should not see the contents written by the subsequent write.`
+  )
+  .params(u =>
+    u //
+      .combine('boundary', kOperationBoundaries)
+      .expand('_context', p => kBoundaryInfo[p.boundary].contexts)
+      .expandWithParams(function* ({ _context }) {
+        for (const readOp of kAllReadOps) {
+          for (const writeOp of kAllWriteOps) {
+            if (checkOpsValidForContext([readOp, writeOp], _context)) {
+              yield {
+                readOp,
+                readContext: _context[0],
+                writeOp,
+                writeContext: _context[1],
+              };
+            }
+          }
+        }
+      })
+  )
+  .fn(async t => {
+    const { readContext, readOp, writeContext, writeOp, boundary } = t.params;
+    const helper = new OperationContextHelper(t);
+
+    const { srcBuffer, dstBuffer } = await t.createBuffersAndTexturesForReadOp(readOp);
+    await t.createBuffersAndTexturesForWriteOp(writeOp, 0, 1);
```

Comment:
here and the rest of the tests, it would be helpful to have comments explaining what's happning, and then also label the numbers by assigning them variables.

IIUC, we:
create srcBuffer and dstBuffer. srcBuffer contents are ??? dstBuffer is 0.

we then read srcBuffer into dstBuffer
and then write value of 1 into srcBuffer.

We check that dstBuffer is still 0.

readValue = 0; // srcBuffer is initially 0 initialized (I think so at least?)
read(srcBuffer, dstBuffer)
write(srcBuffer, someOtherValue);
t.verifyData(dstBuffer, readValue); // got the expected value

so it's kinda a write-read-write test

Reply:
hmm I think that's ok.